distributed_loader: detect highest generation before populating column families

bhalevy opened the pull request:
We should scan all sstables in the table directory and its
subdirectories to determine the highest sstable version and generation
before using it for creating new sstables (via reshard or reshape).

Otherwise, the generations of new sstables created when populating staging (via reshard or reshape) may collide with generations in the base directory, leading to https://github.com/scylladb/scylladb/issues/11789
    
Refs scylladb/scylladb#11789
Fixes scylladb/scylladb#11793

avikivity commented:
> We should

Why?

bhalevy commented:
> > We should
> 
> Why?

"to determine the highest sstable version and generation before using it for creating new sstables"

otherwise, we have the generation of new sstables created when populating staging (via reshard or reshape) may collide with generations in the base directory, leading to https://github.com/scylladb/scylladb/issues/11789

Maybe I should add that to the description of this PR.

bhalevy commented:
Duplicate generation also present an issue for sstable runs, right @raphaelsc?

raphaelsc commented on replica/database.hh:
```diff
@@ -1713,6 +1713,26 @@ public:
+        return get();
+    }
+};
+
```
my preference is that global_column_family_ptr remains in the source file, because it's not used elsewhere and therefore doesn't have to be exposed in a header. we did tons of work around database.hh to reduce compilation time, it should include only essential stuff. even though it's a functionality built upon replica::database, it's only used within a context of sharded sstable loader service. if you really feel like moving it into a header, perhaps we should create database_utils.hh

bhalevy replied:
yes, sorry.  I had a version where sstables_metadata was defined in distributed_loader.hh, requiring global_column_family_ptr, but I reverted that so it's no longer needed. Will remove first 2 patches in the next version

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2717/

raphaelsc commented on replica/distributed_loader.cc:
```diff
@@ -448,57 +428,151 @@ future<> distributed_loader::handle_sstables_pending_delete(sstring pending_dele
-            if (dir_must_exist) {
-                throw std::runtime_error(format("Populating {}/{} failed: {} does not exist", ks, cf, sstdir));
+        for (auto subdir : { "", sstables::staging_dir, sstables::quarantine_dir }) {
+            auto sstpath = get_path(subdir);
```
perhaps we should have start_directory(sstring subdir), where all initialization code goes into it. therefore indentation is kept and start() will simply iterate through subdirs, and call start_directory(...) on each. simplifies reviewing and makes the code easier to follow.

bhalevy replied:
okay

raphaelsc commented on replica/distributed_loader.cc:
```diff
@@ -448,57 +428,151 @@ future<> distributed_loader::handle_sstables_pending_delete(sstring pending_dele
-            return global_table->disable_auto_compaction();
-        }).get();
+    const std::unordered_map<sstring, lw_shared_ptr<sharded<sstables::sstable_directory>>>& directories() const noexcept {
+        return _directories;
```
I feel we need to call it sstable_directories, as directories is too generic.

bhalevy replied:
sure, no problem.
thanks for reviewing

bhalevy commented on replica/distributed_loader.cc:
```diff
@@ -428,57 +428,116 @@ future<> distributed_loader::handle_sstables_pending_delete(sstring pending_dele
+// to reduce the code churn in this transformation
+// TODO: make it a formal class, hiding its private members
+struct sstables_metadata {
+    distributed<replica::database>& db;
```
okay

bhalevy commented:
In v2 ([826fe83](https://github.com/scylladb/scylladb/commit/826fe83e03c3f5cf36309b354c3a5bb864128564)):
- Fixed bug caused by adding an empty subdir to the base fs::path
- Dropped:
  - distributed_loader: move global_column_family_ptr to database.hh
  - database: global_column_family_ptr: use string_view for ks and cf names
  - sstables: delete_atomically: change assertion regarding sstable directory into internal error
- distributed_loader: pre-load all sstables metadata for table before populating it
  - logic squashed with "distrobuted_loader: sstables_metadata: make members private"
  - used async thread to make backportable to older version not supporting coroutines
- (material) new patches:
  - distributed_loader: table_population_metadata: coroutinize start_subdir
  - distributed_loader: coroutinize populate_column_family
    - while at it, since it id modified heavily enough in the series, we might as well turn it into a coroutine and get rid of the async thread it starts for each table sub-directory.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2732/

bhalevy commented:
In v3 ([4d7f0be](https://github.com/scylladb/scylladb/commit/4d7f0be929763ac0c52a0c8e56ce5adc66f0b5fa)):
- distributed_loader: pre-load all sstables metadata for table before populating it
  - print error to find sstable directory in metadata only if exists on disk (should fix CI failures)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2737/

denesb commented:
@raphaelsc please re-review.
